src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java

```diff
@@ -0,0 +1,119 @@
+     */
+    public static final double DUPLICATE_WAY_EPSILON_DEGREES = SphericalDistanceLibrary.metersToDegrees(0.001);
+
+    private final HashGridSpatialIndex<Edge> idx;
+
+    private final LinkingGeoTools linkingGeoTools;
+
+    public EdgesToLinkFinder(HashGridSpatialIndex<Edge> idx, LinkingGeoTools linkingGeoTools) {
+        this.idx = idx;
+        this.linkingGeoTools = linkingGeoTools;
+    }
+
+    /**
+     * finds all near edges that we should link given vertex to
+     */
+    public List<StreetEdge> findEdgesToLink(Vertex vertex, TraverseMode traverseMode) {
+        Envelope env = linkingGeoTools.createEnvelope(vertex);
+
+        List<StreetEdge> candidateEdges = getCandidateEdges(env, traverseMode);
+
+        // Make a map of distances to all edges.
+        TIntDoubleMap distances = getDistances(candidateEdges, vertex);
+
+        sortCandidateEdges(candidateEdges, distances);
+
+        // find the closest candidate edges
+        if (candidateEdges.isEmpty() || distances.get(candidateEdges.get(0).getId()) > RADIUS_DEG) {
+            return emptyList();
+        }
+
+        // find the best edges
+        return getBestEdges(candidateEdges, distances);
+    }
+
+    private List<StreetEdge> getCandidateEdges(Envelope env, TraverseMode traverseMode) {
+        final TraverseModeSet traverseModeSet;
+        if (traverseMode == TraverseMode.BICYCLE) {
+            traverseModeSet = new TraverseModeSet(traverseMode, TraverseMode.WALK);
+        } else {
+            traverseModeSet = new TraverseModeSet(traverseMode);
+        }
+        // We sort the list of candidate edges by distance to the stop
+        // This should remove any issues with things coming out of the spatial index in different orders
+        // Then we link to everything that is within DUPLICATE_WAY_EPSILON_METERS of of the best distance
+        // so that we capture back edges and duplicate ways.
+        return idx.query(env).stream()
+                .filter(StreetEdge.class::isInstance)
+                .map(StreetEdge.class::cast)
+                // note: not filtering by radius here as distance calculation is expensive
+                // we do that below.
+                .filter(edge -> edge.canTraverse(traverseModeSet) &&
+                        // only link to edges still in the graph.
+                        edge.getToVertex().getIncoming().contains(edge))
+                .collect(toList());
+    }
+
+    private TIntDoubleMap getDistances(List<StreetEdge> candidateEdges, Vertex vertex) {
+        TIntDoubleMap distances = new TIntDoubleHashMap();
+        candidateEdges.forEach(e -> distances.put(e.getId(), linkingGeoTools.distance(vertex, e)));
+        return distances;
+    }
+
+    private void sortCandidateEdges(List<StreetEdge> candidateEdges, TIntDoubleMap distances) {
+        // Sort the list.
+        candidateEdges.sort((o1, o2) -> {
```

Comment:
Nie mam pojęcia jak z wydajnością ale lepiej najpierw przemapować na (krawędź, odległość) a później sortować (wiem, że to hashmapa).

Reply:
Zrobione 👍